### PR TITLE
Add support for GitHub Enterprise

### LINKED ISSUE #1129: Allow connection to GitHub Enterprise
The task `cci service connect github` does not allow a user to enter their GitHub enterprise server's address

### PR DESCRIPTION
**Pull request by davidjray**
Fixes #1129 

**Final Implementation Notes**
1. Split GitHub and GitHubEnterprise support into two distinct services. This maintains the historic GitHub service behavior, and adds support for GitHub Enterprise Server via a new service type.
2. In order to support internal Github Enterprise instances we must set an ENV variable `CUMULUSCI_SYSTEM_CERTS` which will allow our connection to support the SSL signed by the internal certificate authority. See [here](https://github.com/SFDO-Tooling/CumulusCI/blob/main/cumulusci/utils/http/requests_utils.py#L23).

Previous work notes:
1. ~~With this implementation the first GitHub service that have a non-github.com domain configured are selected, therefore, this implementation does not support multiple services under a given enterprise "domain", or respect the default service.~~ Fixed by #3271 and #3275
2. ~~The default github.com service is selected instead of the first matching service, which differs from the planned implementation, but this is better.~~ #3271 and #3275
3. ~~We need to refactor the token factory for github services to support the domain supplied when configuring a new service. This means the token must be manually supplied via `--token` when connecting a new github service.~~ Not needed due to #3271
4. ~~Supporting self signed certificates is the default for GitHub Enterprise connections, but this should be optional~~ Considered in #3287 but decided against in favor of setting CUMULUSCI_SYSTEM_CERTS

### CODE REVIEW

File: requirements/prod.in [hunk 1]
```diff
@@ -4,6 +4,7 @@ python-dateutil
 docutils<0.17
 Faker
 fs
+giturlparse
```
Comment (jstvz) [inline requirements/prod.in:7]: Blocked by missing 3PP.

File: requirements/prod.txt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #
-# This file is autogenerated by pip-compile with python 3.8
+# This file is autogenerated by pip-compile with python 3.9
```
Comment (jstvz) [inline requirements/prod.txt:2]: ❔: Mind the version here, pip-tools is platform dependent IIRC.

File: cumulusci/core/github.py [hunk 3]
```diff
@@ -35,9 +38,14 @@
     get_device_code,
     get_device_oauth_token,
 )
+from cumulusci.utils.git import parse_repo_url
 from cumulusci.utils.http.requests_utils import safe_json_from_response
 from cumulusci.utils.yaml.cumulusci_yml import cci_safe_load
 
+# Hiding Warning: Adding certificate verification is strongly advised.
+# See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
+requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
```
Comment (jstvz) [inline cumulusci/core/github.py:47]: 💭: Needs discussion/review.

File: cumulusci/core/github.py [hunk 4]
```diff
@@ -70,11 +78,31 @@ def get_github_api(username=None, password=None):
 INSTALLATIONS = {}
 
 
-def get_github_api_for_repo(keychain, owner, repo, session=None):
-    gh = GitHub(
-        session=session
-        or GitHubSession(default_read_timeout=30, default_connect_timeout=30)
+def _determine_github_client(host: str, client_params: dict) -> GitHub:
```
Comment (davidjray) [inline cumulusci/core/github.py:81]: Note to self... Do we sill like this name?

**Comment by davidjray** [comment]
[f6e5e05](https://github.com/SFDO-Tooling/CumulusCI/pull/3256/commits/f6e5e05fe3b8e822b303519b4d16e63b708be6f8) Fixes some issues I ran into when smoke testing. These fixes were required to overcome issues when configuring the first new "github_enterprise" service, and also access properties of services stored in the keychain.

File: cumulusci/core/github.py [hunk 5]
```diff
@@ -70,11 +72,28 @@ def get_github_api(username=None, password=None):
...
+    # also covers "api.github.com"
+    is_github: bool = host is None or "github.com" in host
+    client_cls: GitHub = GitHub if is_github else GitHubEnterprise  # type: ignore
+    params: dict = client_params
+    if not is_github:
+        params["url"] = "https://" + host  # type: ignore
```
[the remark this answers is not in the document]
Comment (davidjray) [reply cumulusci/core/github.py:81]: Yes the SSL verification will be done. This client, `GitHubEnterprise` has a `verify` flag that can be set to false, but we are not supporting that.

File: cumulusci/core/tests/test_github.py [hunk 6]
```diff
@@ -144,20 +169,96 @@ def test_get_github_api_for_repo__not_installed(self, create_token):
...
+            )
+
+    @responses.activate
+    def test_validate_gh_enterprise(self, keychain_enterprise):
+        keychain_enterprise.set_service(
+            "github_enterprise",
+            "ent2",
+            ServiceConfig(
+                {
+                    "username": "testusername2",
```
[the remark this answers is not in the document]
Comment (davidjray) [reply cumulusci/core/tests/test_github.py:209]: These are NOT real credentials

**Review by jofsky — APPROVED** [review]
Can you ensure setuptools is not a part of prod.txt file?

File: cumulusci/core/github.py [hunk 7]
```diff
@@ -83,33 +101,71 @@ def get_github_api_for_repo(keychain, owner, repo, session=None):
...
+            )
+
+
+def validate_service(options: dict, keychain) -> dict:
     username = options["username"]
     token = options["token"]
-    # Don't make the user wait 4 minutes to fail
-    gh = GitHub(token=token)
+    # Github service doesn't have "repo_domain",
+    repo_domain = options.get("repo_domain", None)
```
[the remark this answers is not in the document]
Comment (jofsky) [reply cumulusci/core/github.py:164]: Can we do the default repo_domain setting for both services?

File: cumulusci/core/config/project_config.py [hunk 8]
```diff
@@ -380,13 +396,11 @@ def repo_commit(self):
                         ):
                             return parts[0]
 
-    def get_github_api(self, owner=None, repo=None):
-        return get_github_api_for_repo(
-            self.keychain, owner or self.repo_owner, repo or self.repo_name
-        )
+    def get_github_api(self, url=None):
```
Comment (jstvz) [inline cumulusci/core/config/project_config.py:399]: Changing this function signature needs to be called out as a breaking change in release notes.

File: cumulusci/core/github.py [hunk 9]
```diff
@@ -70,11 +72,27 @@ def get_github_api(username=None, password=None):
 INSTALLATIONS = {}
 
 
-def get_github_api_for_repo(keychain, owner, repo, session=None):
-    gh = GitHub(
-        session=session
-        or GitHubSession(default_read_timeout=30, default_connect_timeout=30)
+def _determine_github_client(host: str, client_params: dict) -> GitHub:
+    # also covers "api.github.com"
+    is_github: bool = host is None or host == "None" or "github.com" in host
```
Comment (jstvz) [inline cumulusci/core/github.py:77]: ↔️ 🏌🏾‍♂️ 
```suggestion
    is_github: bool = host in (None, "None") or "github.com" in host
```

File: cumulusci/core/github.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jstvz) [reply cumulusci/core/github.py:164]: > Can we do the default repo_domain setting for both services?

Would you say more about why you'd want that?

File: cumulusci/utils/git.py [hunk 10]
```diff
@@ -53,14 +54,33 @@ def construct_release_branch_name(prefix: str, release_identifier: str) -> str:
 
 
 def split_repo_url(url: str) -> Tuple[str, str]:
-    url_parts = url.rstrip("/").split("/")
+    owner, name, _ = parse_repo_url(url)
+    return (owner, name)
+
+
+def parse_repo_url(url: str) -> Tuple[str, str, str]:
+    """Parses a given Github URI into Owner, Repo Name, and Host)
```
Comment (jstvz) [inline cumulusci/utils/git.py:62]:
```suggestion
    """Parses a given Github URI into Owner, Repo Name, and Host
```

**Review by jstvz — APPROVED** [review]

File: cumulusci/cumulusci.yml [hunk 11]
```diff
@@ -1335,6 +1335,22 @@ services:
...
+                required: True
+            email:
+                description: The email address to used by GitHub tasks when an operation requires an email address.
+                required: True
+            repo_domain:
+                description: "GitHub Enterprise domain (Example: git.ent.domain.com)."
+                required: True
+            token:
+                description: Personal Access Token
+                required: True
```
[the remark this answers is not in the document]
Comment (jstvz) [reply cumulusci/cumulusci.yml:1352]: @davidjray Blocking change requested.

**Review by jstvz — COMMENTED** [review]
LGTM

@jofsky I have code in this branch now, so I can't approve. Would you do so when ready?

**Comment by davidmreed** [comment]
@davidjray @jstvz @jofsky I think from above this branch is ready for merge, no? I can go ahead and approve and merge unless anyone has a concern.

**Comment by jstvz** [comment]
@davidmreed: 

> I think from above this branch is ready for merge, no? I can go ahead and approve and merge unless anyone has a concern.

Yeah, looks like it slipped the net when the WI was closed. Please approve.

**Review by prescod — APPROVED** [review]
Approving without review based on David Reed and James' attestation that they've reviewed.

**Pull request merged**